### ISSUE TITLE
Feature Request - ability to copy component tree to clipboard

### ISSUE THREAD

**Issue by robtweed** [facebook/react-devtools#315]
Could you add a mechanism for copying the component tree to the clipboard - either as the expanded tags or (better) a JSON object?  I can't currently find a way to copy the tags from within the Dev Tools window (perhaps I've missed a trick?)

**Comment by spicyj** [comment]
Curious: What would you use this for?

**Comment by robtweed** [comment]
Understandable question, spicyj.  The answer is that I've been looking for a way to graphically visualise a React application's component hierarchy, largely for documentation purposes, but also to allow proper understanding within a team of how the application is constructed and, for example, where a React application's re-rendering events need to be controlled within the component tree.  I've not yet found any reliable tool for creating such a visual representation, but React Developer Tools is one of the few that actually easily and reliably provides the information needed to produce such a visualisation.  The problem is exporting that information from React Dev Tools :-)

**Comment by iamdustan** [comment]
@robtweed would something like https://github.com/algolia/react-element-to-jsx-string#usage solve that for you?

**Comment by robtweed** [comment]
iamdustan - Not sure I see how it would help, but perhaps I'm missing something (which I'm good at! :-)

**Comment by iamdustan** [comment]
If the goal is to `to graphically visualise a React application's component hierarchy`, you can use `react-element-to-jsx` string to get the resulting DOM hierarchy back.

```
console.log(reactElementToJSXString(<My App />);

// would output the following string =>
<div className="root">
  <header className="header">...</header>
  <main className="main">
    ...
  </main>
</div>
```

**Comment by robtweed** [comment]
All I got from using it in my top-level component was
<MainPage
  componentPath={['Top']}
  control={{emit: function noRefCheck() {}, events: {}, handlerFunction: {}, keepAlive: function noRefCheck() {}, log: true, off: function noRefCheck() {}, on: function noRefCheck() {}, sendMessage: function noRefCheck() {}, submitForm: function noRefCheck() {}, updateComponentPath: function noRefCheck() {}}}
/>

No doubt this is due to how and where I added the reactElementToJSXString() command.  My React app is split into individual files - one per component - and browserify is used to compile them together.

Meanwhile React Dev Tools displays the component tree correctly without the need for any invasive code in the application components, but of course it doesn't let me export that info so I can visualise it independently.
